### DISCUSSION TITLE
Pass variables to ScheduleActivityAsync (v3)

### DISCUSSION THREAD

**Discussion by rolandgriesser** [elsa-workflows/elsa-core#4169] [Q&A] [answered]
Hi!

I'm trying to implement a scenario where I have an extended `UserTask` activity that will schedule another activity once the TaskId is generated, so I can e.g. send an email or create a task in an external tool (similar to the External Application example in the documentation, but instead of using a Notification, I'd like to handle that in the workflow).

![image](https://github.com/elsa-workflows/elsa-core/assets/13225475/02e6e253-4ceb-4b6f-bc1f-7163c3cdd705)

The first attempt was to use a Port and in the `ExecuteAsync` set some `Output` properties before calling `ScheduleActivityAsync`. Then in the nested Activity, I wanted to use the new JavaScript syntax `get{Output}From{ActivityId}()`, which was available in the autocomplete but didn't pass the Jint validation.

My current solution is to create a new `ActivityExecutionContext` and add the properties as `Input` values and pass the context to the `ScheduleActivityAsync`:

```
[Port]
        public IActivity? TaskIdGenerated { get; set; }

        protected override async ValueTask ExecuteAsync(ActivityExecutionContext context)
        {
            var identityGenerator = context.GetRequiredService<IIdentityGenerator>();
            var taskId = identityGenerator.GenerateId();
            var httpActivityOptions = context.GetRequiredService<IOptions<HttpActivityOptions>>();
            var callbackUrl = $"{httpActivityOptions.Value.BaseUrl}elsa/api/tasks/{taskId}/complete";

            var payload = new RunTaskBookmarkPayload(taskId, nameof(ExtendedRunTaskActivity));
            context.CreateBookmark(new BookmarkOptions(payload, ResumeAsync, "Elsa.RunTask"));
            context.TransientProperties[BookmarkPropertyKey] = payload;

            if (TaskIdGenerated != null)
            {
                var childContext = context.WorkflowExecutionContext.CreateActivityExecutionContext(TaskIdGenerated);
                childContext.Input.Add("TaskId", taskId);
                childContext.Input.Add("CallbackUrl", callbackUrl);
                await context.ScheduleActivityAsync(TaskIdGenerated, childContext);
            }
        }
```

- Is this the recommended way to pass parameters to a Port?
- Is there also a way to schedule a `FlowNode`? (I'd have to outcomes then, one that will be called at the time of generating the taskId without completing the activity, and one when the activity is completed).

Thank you in advance.

**Comment by sfmskywalker** [comment, marked as answer]
Hello @rolandgriesser,

Thanks for bringing this up. Here's how you can send information to scheduled child activities in Elsa:

When passing information to child activities, use workflow variables. These are automatically available to the child scope. Here's an example:

```csharp
public class ExtendedRunTaskActivity : Activity<string>
{
    [Port] public IActivity? TaskIdGenerated { get; set; }

    protected override async ValueTask ExecuteAsync(ActivityExecutionContext context)
    {
        var taskId = Guid.NewGuid().ToString();

        var scheduleWorkOptions = new ScheduleWorkOptions
        {
            Variables = new List<Variable>
            {
                new Variable<string>("TaskId", taskId)
            },
            CompletionCallback = OnTaskIdGeneratedCompletedAsync
        };

        context.SetResult(taskId);
        await context.ScheduleActivityAsync(TaskIdGenerated, scheduleWorkOptions);
    }

    private async ValueTask OnTaskIdGeneratedCompletedAsync(ActivityCompletedContext context)
    {
        await context.CompleteActivityAsync();
    }
}
```

In this example, the `ScheduleWorkOptions` object includes the generated Task ID, which is then passed when scheduling the child activity. The child activity can now reference the "TaskId" variable using JS (e.g. `getTaskId()`).

On the topic of scheduling a flow node outcome without completing the activity: It's a valid point. The challenge is knowing when a scheduled branch finishes, especially with multiple paths. For linear paths, it's simple. But for branching paths, the flowchart needs a way to recognize the completion of a scheduled branch.

A potential solution is to loop such a branch back to the original activity, connecting to a unique inbound port. This would signal the custom activity not to restart the "Generate Task ID" branch. We're thinking about this functionality, which would allow you to run logic via an outcome without ending the activity.

For now, the best method to schedule activities without ending the parent activity is through "embedded" ports. This makes it easier for the engine to determine when the scheduled child activity is done.

**Reply by rolandgriesser** [reply]
Hi @sfmskywalker,

I think I need your help one more time here as I've been struggling to find a proper solution for this.
I want to be flexible with my UserTasks (e.g. sending an email in some cases, creating a Teams message in others, etc.), so I wanted to use Composite workflows to create those tasks.

Take the example below now: I have a "Webform (Email)" activity. I implemented the Embedded port solution and in the child flow, I call an Http-Endpoint with the UserTask-Id as body, and it returns an url to a webform. This url should then be send via email. But the email body should be supplied via an `Input` to this composite workflow (see the `Body` input on the "Webform (Email)" level).
![image](https://github.com/elsa-workflows/elsa-core/assets/13225475/7196bc1e-1e01-46f4-9fcb-0ee5b5211b83) 
In the Email activity, I supply that value now using the Javascript `getInput` method.
![image](https://github.com/elsa-workflows/elsa-core/assets/13225475/4d464694-a291-420e-b828-1274ca28da1f)
 
In the calling workflow, I now supply the body as an input, and here I can't figure out, how I can make the child activity reevaluate the input again, so I can access the webForm-url returned inside the child activity.
![image](https://github.com/elsa-workflows/elsa-core/assets/13225475/c05dca7f-308a-4cd0-83c0-600d4d76899c)

In the code, I figured out that the parent inputs are passed as variables to the child context, and are evaluated when scheduling already. Is it also somehow possible to reevaluate an expression inside the child? (e.g. if I write a `reevaluateInput` javascript helper?). With the `While` activity, you use `EvaluateInputPropertyAsync`, but I couldn't figure out a way how to apply this inside my javascript helper.

Maybe there is already a solution for what I try to achieve - I'm also thinking about just switching to signal urls - but I like the idea of having an UserTask with an id that I can also track in my external system (like you have in the Elsa guide for employee onboarding).

Thanks in advance!

**Reply by sfmskywalker** [reply]
Hey there, thanks for the detailed description. It's clear what your use case is, and it's a very good one. To achieve this, we will need to make a few adjustments in the engine + studio:

- Similar to what we can do in custom activities, we need to expose the `AutoEvaluate` input property to the Studio so that you can define workflow inputs with that setting, which then instructs the engine to **not** auto-evaluate the input expressions (the default is to auto-evaluate all input expressions).
- We need a JS/Liquid/C#/Python function to let you evaluate input expressions from the designer, and/or provide an activity called e.g. `EvaluateExpression` that let's you do this and provides the result as its output.

Until we have that, your best option is to, indeed, use signaling, or implement custom activities "hard-coded" style, i.e. not using composition - which is obviously less than ideal since it takes away the flexibility you're currently providing.

I'll write up an issue to capture this use case.
